### PR TITLE
Refactor covid scripts

### PR DESCRIPTION
**Pull request by orangejenny**
## Technical Summary
In anticipation of [USH-1678](https://dimagi-dev.atlassian.net/browse/USH-1678), I took a stab at combining Ethan's recent work on bulk case updates with Steph's structure for data cleaning management commands.

There are some changes to what's logged: there are no longer details about **why** a case was skipped, and errors are printed as they're encountered rather than at the end. I think these are acceptable, but feel free to correct me.

Review by commit.

### Safety story
This change is limited to a set of custom scripts that are only run manually.

### Automated test coverage

There's test coverage for each of the management commands.

### QA Plan

No QA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by stephherbers — COMMENTED** [review]
Thank you for the refactoring (and including the run_all_management_command file) I think for the most part it looks great--- I added a couple of comments, but do you mind also linking me to better logger docs?

File: custom/covid/management/commands/update_cases.py [hunk 1]
```diff
@@ -74,5 +82,6 @@ def handle(self, domain, case_type, **options):
         self.extra_options = options
 
         for domain in sorted(domains):
-            print(f"Processing {domain}")
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply custom/covid/management/commands/update_cases.py:69]: You'll still get a note when each domain begins, down on line 73.

**Comment by orangejenny** [comment]
@stephherbers For logger docs, I like [this](https://www.loggly.com/ultimate-guide/python-logging-basics/) guide to the basics. To get a level deeper, the official python docs have an [extensive tutorial](https://docs.python.org/3/howto/logging.html).

File: custom/covid/management/commands/README.rst [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+Most of these scripts build off of the base class ``CaseUpdateCommand``, which has the following features:
+
+* Update all cases of a given type, or supply logic to select a subset
+
+* Update cases in bulk
+
+* Skip cases meeting specific criteria
+
+* Run for either a single domain or a single domain plus all of its linked domains.
+
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply custom/covid/management/commands/README.rst:15]: be52e58fd5b

File: custom/covid/management/commands/clear_owner_ids.py [hunk 3]
```diff
@@ -1,50 +1,17 @@
...
     help = "Makes the owner_id for cases blank"
 
-    def case_block(self, case):
-        blank_owner_id = {'owner_id': '-'}
-        return ElementTree.tostring(CaseBlock(
+    def case_block(self, case, **kwargs):
+        if case.get_case_property('owner_id') == '-':
+            return None
+
+        return [CaseBlock(
```
Comment (esoergel) [inline custom/covid/management/commands/clear_owner_ids.py:16]: ha wow, lotta red in this commit!

File: custom/covid/management/commands/update_owner_ids.py [hunk 4]
```diff
@@ -1,68 +1,45 @@
...
+    def get_location(self, owner_id):
+        if self.locations_objects is None:
+            self.locations_objects = {}
+
+        if owner_id in self.locations_objects:
+            return self.locations_objects[owner_id]
+
+        loc = SQLLocation.objects.get(location_id=owner_id)
+        self.locations_objects[owner_id] = loc
+        return loc
```
Comment (esoergel) [inline custom/covid/management/commands/update_owner_ids.py:27]:
```suggestion
    @memoized
    def get_location(self, owner_id):
        return SQLLocation.objects.get(location_id=owner_id)
```

File: custom/covid/management/commands/update_cases.py [hunk 5]
```diff
@@ -13,58 +18,66 @@ class CaseUpdateCommand(BaseCommand):
...
         parser.add_argument('--and-linked', action='store_true', default=False)
-        parser.add_argument('--output-file', type=str, default=None)
+        parser.add_argument('--throttle-secs', type=float, default=0)
 
     def handle(self, domain, case_type, **options):
+        # logger.debug will record something to a file but not print it
+        self.logger = logging.getLogger(self.logger_name())
+        if not settings.UNIT_TESTING:
+            self.logger.addHandler(logging.FileHandler(self.logger_name().split(".")[-1] + ".txt"))
+        self.logger.setLevel(logging.DEBUG)
```
Comment (esoergel) [inline custom/covid/management/commands/update_cases.py:50]: you could also make `self.logger` a `@cached_property` to help keep `handle` clean

File: custom/covid/management/commands/update_cases.py [hunk 6]
```diff
@@ -13,58 +18,66 @@ class CaseUpdateCommand(BaseCommand):
...
 
-        options.pop("and_linked")
-        options.pop("username")
-        options.pop("output_file", None)
         self.extra_options = options
 
-        for domain in sorted(domains):
-            print(f"Processing {domain}")
-            self.update_cases(domain, case_type, user_id)
+        username = user_id_to_username(user_id)
```
Comment (esoergel) [inline custom/covid/management/commands/update_cases.py:70]: You can delete this line and all the user_id stuff above.  The dev running the command passes in a username, and we convert it to a user ID and then back to a username here. `SystemFormMeta.for_script` does the validation as well.

File: custom/covid/management/commands/update_cases.py [hunk 7]
```diff
@@ -13,58 +18,66 @@ class CaseUpdateCommand(BaseCommand):
     """
 
     def __init__(self):
-        self.output_file = None
         self.extra_options = {}
 
-    def case_block(self):
+    def logger_name(self):
```
Comment (esoergel) [inline custom/covid/management/commands/update_cases.py:23]: If you make this a `@property`, you can still override it with a simple `logger_name = __name__` in the subclasses:
```python
In [1]: class Foo:
   ...:     @property
   ...:     def name(self):
   ...:         raise NotImplementedError()

In [2]: class Bar(Foo):
   ...:     name = "hello"

In [3]: Bar().name
Out[3]: 'hello'
```

File: custom/covid/management/commands/update_cases.py [hunk 8]
```diff
@@ -44,7 +45,8 @@ def add_arguments(self, parser):
     def handle(self, domain, case_type, **options):
         # logger.debug will record something to a file but not print it
         self.logger = logging.getLogger(self.logger_name())
-        self.logger.addHandler(logging.FileHandler(self.logger_name().split(".")[-1] + ".txt"))
+        if not settings.UNIT_TESTING:
```
Comment (esoergel) [inline custom/covid/management/commands/update_cases.py:48]: :+1:

**Review by esoergel — APPROVED** [review]

File: custom/covid/management/commands/update_cases.py [hunk 9]
```diff
@@ -45,6 +47,7 @@ def add_arguments(self, parser):
         parser.add_argument('--username', type=str, default=None)
         parser.add_argument('--and-linked', action='store_true', default=False)
         parser.add_argument('--output-file', type=str, default=None)
+        parser.add_argument('--throttle-secs', type=float, default=0)
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply custom/covid/management/commands/update_cases.py]: 👍 Removed in 60ad6a5212e - do you think it's worth keeping in the `hqcase.bulk` code?

File: custom/covid/management/commands/update_case_index_relationship.py [hunk 10]
```diff
@@ -41,41 +33,36 @@ def get_owner_id(case_type):
...
+        This command has an optional argument '--location' that will exclude all cases with that location. If the
+        case_type is lab_result, the owner_id of that extension case is set to '-'.
+    """
+
+    def logger_name(self):
+        return __name__
+
+    def find_case_ids(self, domain):
+        return CommCareCase.objects.get_case_ids_in_domain(domain, self.case_type)
+
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply custom/covid/management/commands/update_case_index_relationship.py]: suggestion didn't work for some reason - 70cfed2f456

File: custom/covid/management/commands/update_owner_ids.py (hunk 4 above)
Comment (orangejenny) [reply custom/covid/management/commands/update_owner_ids.py:27]: ha, good point

File: custom/covid/management/commands/add_assignment_cases.py [hunk 11]
```diff
@@ -33,59 +33,52 @@ def find_owner_id(case, case_property):
...
     help = "Creates assignment cases for cases of a specified type in an active location"
 
-    def case_block(self, case, owner_id, assignment_type):
+    def logger_name(self):
+        return __name__
+
+    def _case_block(self, case, owner_id, assignment_type):
         case_id = uuid.uuid4().hex
-        return ElementTree.tostring(CaseBlock.deprecated_init(
+        return CaseBlock.deprecated_init(
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply custom/covid/management/commands/add_assignment_cases.py:41]: 631e7f10df2

File: custom/covid/management/commands/update_cases.py (hunk 7 above)
Comment (orangejenny) [reply custom/covid/management/commands/update_cases.py:24]: nice, I didn't know that - 6982c65ad1ea22ffb6534ef66ab977289698d360

File: custom/covid/management/commands/update_cases.py [hunk 12]
```diff
@@ -55,27 +55,25 @@ def add_arguments(self, parser):
         parser.add_argument('--throttle-secs', type=float, default=0)
 
     def handle(self, domain, case_type, **options):
+        print(f"{datetime.datetime.utcnow()} Starting run: {options}")
         domains = {domain}
-        if options["and_linked"]:
+        if options.pop("and_linked", False):
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply custom/covid/management/commands/update_cases.py]: ohhhh, that makes sense, I forgot that. I had been aggravated that `options` doesn't have a `discard` function, but now that makes sense - popping a known argument is never going to throw a `KeyError`. Updated in 8b3f28a93d9

File: custom/covid/management/commands/update_cases.py (hunk 6 above)
Comment (orangejenny) [reply custom/covid/management/commands/update_cases.py]: 42d81f50469

File: custom/covid/management/commands/update_cases.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply custom/covid/management/commands/update_cases.py]: Not particularly

File: custom/covid/management/commands/update_owner_ids.py [hunk 13]
```diff
@@ -1,68 +1,39 @@
...
-
-        total = 0
-        for chunk in chunked(case_blocks, BATCH_SIZE):
-            submit_case_blocks(chunk, domain, device_id=DEVICE_ID, user_id=user_id)
-            total += len(chunk)
-            print("Updated {} cases on domain {}".format(total, domain))
-
-        self.log_data(domain, "update_owner_ids", case_type, len(case_ids), total, errors)
+    locations_objects = None
+
```
Comment (esoergel) [inline custom/covid/management/commands/update_owner_ids.py:16]:
```suggestion
```
This isn't needed anymore

File: custom/covid/management/commands/update_cases.py [hunk 14]
```diff
@@ -13,58 +17,55 @@ class CaseUpdateCommand(BaseCommand):
...
     def add_arguments(self, parser):
         parser.add_argument('domain')
         parser.add_argument('case_type')
         parser.add_argument('--username', type=str, default=None)
         parser.add_argument('--and-linked', action='store_true', default=False)
-        parser.add_argument('--output-file', type=str, default=None)
 
     def handle(self, domain, case_type, **options):
+        # logger.debug will record something to a file but not print it
+        self.logger = logging.getLogger(self.logger_name())
```
Comment (esoergel) [inline custom/covid/management/commands/update_cases.py:46]: these are properties now
```suggestion
        self.logger = logging.getLogger(self.logger_name)
```
(a couple references below this one too)

File: custom/covid/management/commands/update_cases.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply custom/covid/management/commands/update_cases.py]: ☠️ b96a0b0936a

File: custom/covid/management/commands/update_cases.py (hunk 14 above)
Comment (orangejenny) [reply custom/covid/management/commands/update_cases.py]: [2a7caeb](https://github.com/dimagi/commcare-hq/pull/31168/commits/2a7caebf54ac1d4a374d538b6cd1a94f28078cad)

**Review by esoergel — APPROVED** [review]

**Pull request merged**